hudi-spark-datasource/hudi-spark3-common/src/main/scala/org/apache/spark/sql/hudi/catalog/HoodieCatalog.scala

```diff
@@ -153,6 +155,72 @@ class HoodieCatalog extends DelegatingCatalogExtension
     }
   }
 
+  override def loadTable(ident: Identifier, timestamp: Long): Table = {
+    super.loadTable(ident) match {
+      case V1Table(catalogTable0) if sparkAdapter.isHoodieTable(catalogTable0) =>
+
+        // spark passes microseconds
+        val milliseconds = TimeUnit.MICROSECONDS.toMillis(timestamp)
+        val targetTimestamp = HoodieInstantTimeGenerator.formatMillis(milliseconds)
+
+        constructTable(ident, catalogTable0, targetTimestamp)
+
+      case t => t
+    }
+  }
+
+  override def loadTable(ident: Identifier, version: String): Table = {
+    super.loadTable(ident) match {
+      case V1Table(catalogTable0) if sparkAdapter.isHoodieTable(catalogTable0) =>
+
+        if (!HoodieInstantTimeGenerator.isValidInstantTime(version)) {
+          throw new AnalysisException(s"invalid snapshot id: $version")
+        }
+
+        constructTable(ident, catalogTable0, version);
+
+      case t => t
+    }
+  }
+
+  private def constructTable(ident: Identifier,
+                             catalogTable0: CatalogTable,
+                             timestampAsOf: String): Table = {
+    val catalogTable = catalogTable0.comment match {
+      case Some(v) =>
+        val newProps = catalogTable0.storage.properties + (
```

Comment:
catalogTable0.storage.copy(properties = newProps)) is this creating a new instance so the original catalog table is not modified.